src/uct/sm/scopy/base/scopy_iface.c

```diff
@@ -55,17 +74,70 @@ UCS_CLASS_INIT_FUNC(uct_scopy_iface_t, uct_scopy_iface_ops_t *ops, uct_md_h md,
 {
     uct_scopy_iface_config_t *config = ucs_derived_of(tl_config,
                                                       uct_scopy_iface_config_t);
+    ucs_status_t status;
 
     UCS_CLASS_CALL_SUPER_INIT(uct_sm_iface_t, &ops->super, md, worker, params, tl_config);
 
-    self->config.max_iov = ucs_min(config->max_iov, ucs_iov_get_max());
+    self->tx              = ops->ep_tx;
+    self->config.max_iov  = ucs_min(config->max_iov, ucs_iov_get_max());
+    self->config.seg_size = ((config->seg_size == UCS_MEMUNITS_AUTO) ?
+                             UCT_SCOPY_IFACE_SEG_SIZE :
+                             config->seg_size);
 
-    return UCS_OK;
+    UCS_STATIC_ASSERT(sizeof(uct_scopy_comp_t) <= sizeof(uct_scopy_tx_t));
+
+    if (config->tx_mpool.bufs_grow == 0) {
+        ucs_error("TX mpool the buffers grow value must be > 0");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (config->tx_mpool.max_bufs == 0) {
+        ucs_error("TX mpool the maximal number of buffers must be > 0");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    ucs_arbiter_init(&self->arbiter);
+
+    status = ucs_mpool_init(&self->tx_mpool, 0,
+                            sizeof(uct_scopy_tx_t) +
+                            self->config.max_iov * sizeof(uct_iov_t),
+                            0, UCS_SYS_CACHE_LINE_SIZE,
+                            config->tx_mpool.bufs_grow,
+                            config->tx_mpool.max_bufs,
+                            &uct_scopy_mpool_ops, "uct_scopy_iface_tx_mp");
+
+    return status;
 }
 
 static UCS_CLASS_CLEANUP_FUNC(uct_scopy_iface_t)
 {
-    /* No op */
+    self->super.super.super.ops.iface_progress_disable(&self->super.super.super,
+                                                       UCT_PROGRESS_SEND |
+                                                       UCT_PROGRESS_RECV);
+    ucs_mpool_cleanup(&self->tx_mpool, 1);
+    ucs_arbiter_cleanup(&self->arbiter);
 }
 
 UCS_CLASS_DEFINE(uct_scopy_iface_t, uct_sm_iface_t);
+
+unsigned uct_scopy_iface_progress(uct_iface_h tl_iface)
+{
+    uct_scopy_iface_t *iface = ucs_derived_of(tl_iface, uct_scopy_iface_t);
+    unsigned count           = 0;
+    ucs_arbiter_dispatch(&iface->arbiter, 1, uct_scopy_ep_progress_tx, &count);
+    return count;
+}
+
+ucs_status_t uct_scopy_iface_flush(uct_iface_h tl_iface, unsigned flags,
+                                   uct_completion_t *comp)
+{
+    uct_scopy_iface_t *iface = ucs_derived_of(tl_iface, uct_scopy_iface_t);
+
+    if (!ucs_arbiter_is_empty(&iface->arbiter)) {
+        UCT_TL_IFACE_STAT_FLUSH_WAIT(&iface->super.super);
+        return uct_scopy_iface_handle_flush_wait(iface, iface->last_tx, comp);
```

Comment:
1. where is this `last_tx` assigned?
2. It should track state of all endpoints, not a single one where `last` operation was executed